hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastTypeCoercion.java

```diff
@@ -0,0 +1,422 @@
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlUtil;
+import org.apache.calcite.sql.fun.SqlCase;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.SqlTypeFamily;
+import org.apache.calcite.sql.type.SqlTypeUtil;
+import org.apache.calcite.sql.validate.SqlValidatorScope;
+import org.apache.calcite.sql.validate.implicit.TypeCoercionImpl;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.hazelcast.sql.impl.calcite.validate.SqlNodeUtil.isLiteral;
+import static com.hazelcast.sql.impl.calcite.validate.SqlNodeUtil.isParameter;
+import static com.hazelcast.sql.impl.calcite.validate.SqlNodeUtil.numericValue;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.canCast;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.isChar;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.isFloatingPoint;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.isInteger;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.isNumeric;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.isTemporal;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.narrowestTypeFor;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.typeName;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.withHigherPrecedence;
+import static com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeSystem.withHigherPrecedenceForLiterals;
+import static org.apache.calcite.sql.SqlKind.BETWEEN;
+import static org.apache.calcite.sql.SqlKind.BINARY_ARITHMETIC;
+import static org.apache.calcite.sql.SqlKind.BINARY_COMPARISON;
+import static org.apache.calcite.sql.SqlKind.BINARY_EQUALITY;
+import static org.apache.calcite.sql.SqlKind.MINUS_PREFIX;
+import static org.apache.calcite.sql.SqlKind.PLUS_PREFIX;
+import static org.apache.calcite.sql.type.SqlTypeName.ANY;
+import static org.apache.calcite.sql.type.SqlTypeName.BIGINT;
+import static org.apache.calcite.sql.type.SqlTypeName.BOOLEAN;
+import static org.apache.calcite.sql.type.SqlTypeName.CHAR_TYPES;
+import static org.apache.calcite.sql.type.SqlTypeName.DECIMAL;
+import static org.apache.calcite.sql.type.SqlTypeName.DOUBLE;
+import static org.apache.calcite.sql.type.SqlTypeName.NULL;
+
+/**
+ * Provides custom coercion strategies supporting {@link HazelcastIntegerType}
+ * and assigning more precise types comparing to the standard Calcite coercion.
+ */
+public final class HazelcastTypeCoercion extends TypeCoercionImpl {
```

Comment:
There is one strange thing about type coercion I observe with IN and OR operators. Consider that we have a table `t(f INT)`, The following two queries are presumably equivalent:
```
SELECT * FROM t WHERE f=? OR f=?
SELECT * FROM t WHERE f IN (?,?)
```
However, generated filters are not equivalent:
```
OR(=(CAST($0):BIGINT(63), ?0), =(CAST($0):BIGINT(63), ?0))
OR(=($0, ?0), =($0, ?1))
```
That is, we do upcast for `OR`, but not for `IN`.

Reply:
The same problem applies to `BETWEEN` statement: there is coercion for 'f >= ? AND f <= ?', but not for `f BETWEEN ? AND ?`.

Reply:
There is no support for `in` and `between` in this PR. I started some work on supporting them in `sql` branch, but decided to postpone it until basic expressions would be working. As far as I remember, the conversion of `in` into `or` and `between` into `and` is done during generation of `RexNode`s, but the coercion works on `SqlNode` level before the `RexNode`s are generated, that's why the outputs are not equivalent.

Reply:
Got it. Can we have a ticket for 4.2 for this, to ensure that we do not miss this fix?

Reply:
Thinking more about it, I have one additional question - what would be the behavior of the query `SELECT * FROM t WHERE f_int IN (?, ?)`, when the passed parameter is `Long`? Will it fail on the parameter validation stage or during actual query execution? What would be the error message?

Reply:
At the current state of this PR, it should fail during parameter validation if the passed `long` value is outside of the `int` representable range, the message should mention the parameter position with a failure reason specified as something like "numeric overflow ...".

Reply:
Created an issue https://github.com/hazelcast/hazelcast/issues/17303 for 4.2, but there is still a chance it can be squeezed to 4.1.